src/ecdh.rs

```diff
@@ -182,6 +189,68 @@ impl SharedSecret {
         y.copy_from_slice(&xy[32..]);
         hash_function(x, y)
     }
+
+    /// Formats the explicit byte value of the shared secret kept inside the type as a
+    /// little-endian hexadecimal string using the provided formatter.
+    ///
+    /// This is the only method that outputs the actual shared secret value, and, thus,
+    /// should be used with extreme precaution.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[cfg(all(feature = "std", not(feature = "bitcoin_hashes")))] {
+    /// # use std::str::FromStr;
+    /// # use secp256k1::{SecretKey, PublicKey};
+    /// use secp256k1::ecdh::SharedSecret;
+    /// # let pk = PublicKey::from_slice(&[3, 23, 183, 225, 206, 31, 159, 148, 195, 42, 67, 115, 146, 41, 248, 140, 11, 3, 51, 41, 111, 180, 110, 143, 114, 134, 88, 73, 198, 174, 52, 184, 78]).expect("hard coded slice should parse correctly");
+    /// # let sk = SecretKey::from_str("57f0148f94d13095cfda539d0da0d1541304b678d8b36e243980aab4e1b7cead").unwrap();
+    /// let secret = SharedSecret::new(&pk, &sk);
+    /// // Normal display hides the value.
+    /// assert_eq!(format!("{:?}", secret), "SharedSecret(#6b778f6da1cd7d46)");
+    /// // Here we explicitly display the secret value:
+    /// assert_eq!(
+    ///     format!("{}", secret.display_secret()),
+    ///     "cf05ae7da039ddce6d56dd57d3000c6dd91c6f1695eae47e05389f11e24670430000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000");
+    /// // Can explicitly display with `{}` and `{:?}`.
+    /// assert_eq!(
+    ///     format!("{:?}", secret.display_secret()),
+    ///     format!("DisplaySecret(\"{}\")", secret.display_secret()));
+    /// # }
+    /// ```
+    #[inline]
+    pub fn display_secret(&self) -> DisplaySecret {
+        DisplaySecret { secret: self.serialize_secret() }
+    }
+}
+
+/// Helper struct for safely printing a shared secret.
+///
+/// Formats the explicit byte value of the secret as a little-endian hexadecimal string using the
+/// provided formatter.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct DisplaySecret {
+    secret: [u8; BUF_SIZE]
+}
```

Comment:
You can't derive these because const generics are unsupported in older Rust versions and arrays have the traits implemented only up to size of 32.

I think those traits are somewhat strange anyway but if you really want to you can impl them manually.
Also this looks like a good case for `ArrayVec` in the future.

OTOH, I think reference is better. It fixes this, bug below and the performance improvement may be interesting. Lifetime is non-issue as the wrapper is not supposed to be stored - just call passed to some formatting macro.